File: c-src/cmd.c
#include "cmd.h"

#define MAXBIND 15
#define MAXKEYLENGTH 25
#define MAXCMDLENGTH 25

/* Keybinding pair. Pairs an integer key value
	with the command it is bound to */
typedef struct {
	unsigned int key;
	Command cmd;
} Key_map;

/* Key bindings. Initialized to default values */
static Key_map Key_bindings[MAXBIND] = {
	{KEY_LEFT,  CMD_LEFT},
	{KEY_RIGHT, CMD_RIGHT},
	{KEY_UP,    CMD_UP},
	{KEY_DOWN,  CMD_DOWN},
	{10,        CMD_ACCEPT},
	{27,        CMD_CANCEL},
	{'q',       CMD_QUIT},
	{'s',       CMD_SAVE},
	{'f',       CMD_FIRE},
	{'i',       CMD_INVENTORY},
	{'u',       CMD_USE},

	/* Vikeys */
	{'k',       CMD_UP},
	{'j',       CMD_DOWN},
	{'h',       CMD_LEFT},
	{'l',       CMD_RIGHT},
};

/*-------------------Config file definitions */
struct key_pair {
	char str[MAXKEYLENGTH];
	unsigned int key;
};
struct cmd_pair {
	char str[MAXCMDLENGTH];
	Command cmd;
};

static struct key_pair key_pairs[] = {
	{"arrow_left",  KEY_LEFT},
	{"arrow_right", KEY_RIGHT},
	{"arrow_up",    KEY_UP},
	{"arrow_down",  KEY_DOWN},
	{"escape",      27},
	{"enter",       10},
	{"q",           'q'},
	{"s",           's'},
	{"f",           'f'},
	{"i",           'i'},
	{"u",           'u'},
};
static int numkeypair = sizeof(key_pairs)/sizeof(struct key_pair);

static struct cmd_pair cmd_pairs[] = {
	{"left", 				CMD_LEFT},
	{"right",				CMD_RIGHT},
	{"up",					CMD_UP},
	{"down",				CMD_DOWN},
	{"save",				CMD_SAVE},
	{"quit",				CMD_QUIT},
	{"target/fire",	CMD_FIRE},
	{"inventory",		CMD_INVENTORY},
	{"accept",			CMD_ACCEPT},
	{"use",					CMD_USE},
	{"cancel",			CMD_CANCEL},
};
static int numcmdpair = sizeof(cmd_pairs)/sizeof(struct key_pair);
/*-------------------------------------------*/

/* Takes a keyboard input and returns the command
	that input is bound to in Key_bindings */
Command lookup_cmd(unsigned int key) {
	int i;

	for (i=0; i<MAXBIND; i++) {
		if (Key_bindings[i].key == key) {
			return Key_bindings[i].cmd;
		}
	}

	return CMD_NONE;
}

/* Simple get string function
	Can you tell I hate input handling yet? */
static void get_string(char str[], FILE *fp) {
	int ch;
	int i = 0;

	ch = getc(fp);
	while(isspace(ch)) {
		ch = getc(fp);
	}

	while(!isspace(ch)) {
		str[i++] = ch;
		ch = getc(fp);
	}
	str[i] = '\0';
}


/* Reads the keybindings defined in the file fp.
	The keybindings should start where the cursor
	currently lies and end at an [end] string */
int load_bindings(FILE *fp) {
	char str1[MAXKEYLENGTH], str2[MAXCMDLENGTH], ch;
	unsigned int key, i;
	Command cmd;

	get_string(str1, fp);
	while(strcmp(str1, "[end]") != 0) {
		get_string(str2, fp);

		/* Hackey - translate the config key to the ascii key */
		for (i=0; i<numkeypair; i++) {
			if (strncmp(str1, key_pairs[i].str, strlen(key_pairs[i].str)) == 0) {
				key = key_pairs[i].key;
				i = numkeypair;
			}
		}

		/* Hackey again - translate the config command to game command */
		for (i=0; i<numcmdpair; i++) {
			if (strncmp(str2, cmd_pairs[i].str, strlen(cmd_pairs[i].str)) == 0) {
				cmd = cmd_pairs[i].cmd;
				i = numcmdpair;
			}
		}

		/* Search the the list of keybindings and replace any binding
			to the given key with the new binding */
		for (i=0; i<MAXBIND; i++) {
			if (Key_bindings[i].key == key || Key_bindings[i].key == 0) {
				Key_bindings[i].cmd = cmd;
				i = MAXBIND;
			}
		}

		get_string(str1, fp);
	}

	return 0;
}

File: src/cmd.ml
open Scanf;;
open Curses;;

(* Commands in the internal representation *)
type command =
  | None
  | Left
  | Right
  | Up
  | Down
  | Accept
  | Cancel
  | Quit
  | Save
  | Attack
  | Inventory
  | Use

module HT = Hashtbl.Make(struct
                           type t = int
                           let equal = (=)
                           let hash = Hashtbl.hash
                         end)
(* Stores the set of key mappings *)
let key_map = 
  (* Default keybindings, incase the config file is missing *)
  let defaults = [
    (Key.left, Left);
    (Key.right, Right);
    (Key.up, Up);
    (Key.down, Down);
    (10, Accept);
    (27, Cancel);
    (int_of_char 'q', Quit);
    (int_of_char 's', Save);
    (int_of_char 'f', Attack);
    (int_of_char 'i', Inventory);
    (int_of_char 'u', Use);

    (* Vikeys *)
    (int_of_char 'k', Up);
    (int_of_char 'j', Down);
    (int_of_char 'h', Left);
    (int_of_char 'l', Right)]
  in
	let map = HT.create 32 in
    List.iter (fun (key, cmd) -> HT.add map key cmd) defaults;
		map
              
(* Lookup a keypress from curses "getch" *)
let lookup key =
  try HT.find key_map key
  with Not_found -> None

let lookup_key str = match str with
  | "arrow_left"  -> Key.left
	| "arrow_right" -> Key.right
	| "arrow_up"    -> Key.up
	| "arrow_down"  -> Key.down
	| "escape"      -> 27
	| "enter"       -> 10
	| _             -> 
	  if String.length str > 1 then
		  raise Not_found
		else 
	    int_of_char (str.[0])

let lookup_cmd str = match str with
  | "left" -> Left
	| "right" -> Right
	| "up"    -> Up
	| "down"  -> Down
	| "save"   -> Save
	| "quit"   -> Quit
	| "attack" -> Attack
	| "use"    -> Use
	| "inventory" -> Inventory
	| "accept"    -> Accept
	| "cancel"    -> Cancel
	| _           -> raise Not_found


(* Loads the keybindings *)
let load_bindings ic =
  let rec parse_pair str1 str2 = match str2 with
	  | "[end]" | "" -> ()
		| _ -> begin
		    Log.debug ("Cmd: Binding '" ^ str1 ^ "' to '" ^ str2 ^ "'");
				begin
          try HT.add key_map (lookup_key str1) (lookup_cmd str2)
          with Not_found -> ()
			  end;
        fscanf ic " %s " parse_string
      end
  and parse_string str = match str with
	  | "[end]" | "" -> ()
		| _ -> fscanf ic " %s " (parse_pair str)
	in
	  fscanf ic " %s " parse_string
